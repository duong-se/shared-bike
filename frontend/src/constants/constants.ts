export const tokenKey = 'accessToken'

export const commonHeaders = {
  'Accept': 'application/json',
  'Content-Type': 'application/json',
}
