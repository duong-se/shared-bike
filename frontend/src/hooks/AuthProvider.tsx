import React, { useEffect } from "react"
import jwtDecode from 'jwt-decode'
import { tokenKey } from "../constants/constants";
import { User } from "../typings/types";


type AuthContextType = {
  setUser: React.Dispatch<React.SetStateAction<User | undefined>>
  user?: User
}

const AuthContext = React.createContext<AuthContextType>(null!);

export const useAuth = () => {
  return React.useContext(AuthContext);
}

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [user, setUser] = React.useState<User|undefined>()
  useEffect(() => {
    const token = localStorage.getItem(tokenKey)
    if (token) {
      const decoded = jwtDecode<User>(token);
      setUser(decoded);
    }
  }, [])
  return <AuthContext.Provider value={{ user, setUser }}>{children}</AuthContext.Provider>;
}
