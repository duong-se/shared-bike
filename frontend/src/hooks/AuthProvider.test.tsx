import { render, screen, waitFor } from '@testing-library/react'
import { tokenKey } from '../constants/constants'
import { AuthProvider, useAuth } from './AuthProvider'

describe('AuthProvider', () => {
  it('should run correctly and show user', async () => {
    const MockComponent = () => {
      const { user } = useAuth()
      return <div>{user?.name}</div>
    }
    localStorage.setItem(tokenKey, 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJ0ZXN0MSIsIm5hbWUiOiJUZXN0IFVzZXIiLCJwZXJtaXNzaW9ucyI6bnVsbCwiZXhwIjoxNjU3MzUyODM0fQ.jU5yp2y-3H-dxXP1hdDW-FYEYv5F9GhAVDCbafphUzI')
    render(<AuthProvider><MockComponent/></AuthProvider>)
    const user = await screen.findByText('Test User')
    await waitFor(() => {
      expect(user).toBeInTheDocument()
    })
  })

  it('should not show user', async () => {
    const MockComponent = () => {
      const { user } = useAuth()
      return <div>{user?.name}</div>
    }
    localStorage.clear()
    const { container } = render(<AuthProvider><MockComponent/></AuthProvider>)
    expect(container).toMatchSnapshot()
  })
})
