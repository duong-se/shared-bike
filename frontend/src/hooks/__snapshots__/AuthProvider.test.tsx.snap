// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider should not show user 1`] = `
<div>
  <div />
</div>
`;
